Add integration for Belgian weather provider meteo.be

jdejaegh opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

This PR adds an integration to support the [Royal Meteorological Institute of Belgium](https://www.meteo.be) as a weather provider. 
The code is from a custom component I wrote: https://github.com/jdejaegh/irm-kmi-ha

This implementation only contains code for the weather platform to keep the PR as small as possible. 
I plan to submit subsequent PR once this one is merged.  The goal is to reach the same features as in the custom component.

Code is documented with doc string, comments and tested with 17 unit tests.  I'll look at how to add documentation for users in the coming days and open a PR for that as well.

This is my first contribution to Home Assistant Core so please point to resources I may have missed :slightly_smiling_face: 

Note about the +7,677 lines: this is not all Python code, most of it are JSON test fixtures.  There are 838 lines of actual component code and 613 lines of tests.
## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- Link to documentation pull request: home-assistant/home-assistant.io#38978
- Link to dependency:  https://github.com/jdejaegh/irm-kmi-api
- Link to custom component:  https://github.com/jdejaegh/irm-kmi-ha

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] I have followed the [perfect PR recommendations][perfect-pr]
- [X] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

jdejaegh committed:
Add IRM KMI integration code

Ignoring quality scale linting for now

jdejaegh committed:
Add strings for IRM KMI

jdejaegh committed:
Use runtime data

jdejaegh committed:
Update generated json

jdejaegh committed:
Add repairs

jdejaegh committed:
Use config entry runtime data

jdejaegh committed:
Add tests for the IRM KMI integration

jdejaegh committed:
Update quality_scale.yaml

silamon commented on homeassistant/components/irm_kmi/__init__.py:
```diff
@@ -0,0 +1,64 @@
+        cdt_map=IRM_KMI_TO_HA_CONDITION_MAP,
+    )
+
+    hass.data[DOMAIN][entry.entry_id] = irm_kmi_coordinator = IrmKmiCoordinator(
```
If possible, make a dedicated class where you keep the client and the coordinator. Then use the dedicated class and move that completely in runtime data.

jdejaegh replied:
I made a class as suggested.  However, I had to also give the API client to the IrmKmiCoordinator and not just put it in the ConfigEntry or I got circular dependencies.

silamon commented on homeassistant/components/irm_kmi/__init__.py:
```diff
@@ -0,0 +1,64 @@
+        hass, entry
+    )
+
+    # When integration is set up, set the logging level of the irm_kmi_api package to the same level to help debugging.
+    logging.getLogger("irm_kmi_api").setLevel(_LOGGER.getEffectiveLevel())
```
You can add the loggers to the manifest
```"loggers": ["irm_kmi_api"],```
Then this code is not needed.

silamon commented on homeassistant/components/irm_kmi/__init__.py:
```diff
@@ -0,0 +1,64 @@
+    await hass.config_entries.async_reload(entry.entry_id)
+
+
+async def async_migrate_entry(
+    _hass: HomeAssistant, _config_entry: IrmKmiConfigEntry
+) -> bool:
+    """Migrate old entry."""
+    # There are no previous versions of the config entry at this time.
+    return True
```
This code shouldn't be needed. Since we are not migrating anything yet.

silamon commented on homeassistant/components/irm_kmi/coordinator.py:
```diff
@@ -0,0 +1,134 @@
+        :return: ProcessedCoordinatorData
+        """
+        # When integration is set up, set the logging level of the irm_kmi_api package to the same level to help debugging.
+        logging.getLogger("irm_kmi_api").setLevel(_LOGGER.getEffectiveLevel())
```
Can also be removed when loggers is set

jdejaegh committed:
Update quality scale check list after PR for documentation

jdejaegh committed:
Update CODEOWNER and requirements

NoRi2909 commented on homeassistant/components/irm_kmi/strings.json:
```diff
@@ -0,0 +1,70 @@
+        "step": {
+          "confirm": {
+            "title": "Repair: {zone} is outside of Benelux",
+            "description": "This integration can only get data for location in the Benelux. Move the zone or delete this configuration entry."
```
Better grammar: `"This integration can only get data for locations in Benelux. …"`

jdejaegh replied:
Thanks for the grammar hints: English is not my native language.  I fixed the three errors in commit 5f875bf95068ff086baf411733b28a1820c2ab20

NoRi2909 commented on homeassistant/components/irm_kmi/strings.json:
```diff
@@ -0,0 +1,70 @@
+          }
+        },
+        "error": {
+          "out_of_benelux": "{zone} is out of Benelux.  Move it inside Benelux first.",
```
Use "outside" like in lines 53/57.
Remove duplicated space character.

-> `"{zone} is outside of Benelux. Move it inside Benelux first."`

NoRi2909 commented on homeassistant/components/irm_kmi/strings.json:
```diff
@@ -0,0 +1,70 @@
+      }
+    },
+    "error": {
+      "out_of_benelux": "{zone} is out of Benelux.  Pick a zone in Benelux.",
```
Use "outside" like in lines 53/57.
Remove duplicated space character.

-> `"{zone} is outside of Benelux. Pick a zone in Benelux."`

jdejaegh committed:
Update strings else tests somehow fail

jdejaegh committed:
Use a data class for ConfigEntry.runtime_data

jdejaegh committed:
Use manifest to set loggers

jdejaegh committed:
Remove unused code

jdejaegh committed:
Improve wording in strings.json

jdejaegh committed:
Add _attr_has_entity_name = True

jdejaegh committed:
bump irm-kmi-api==1.0.3

silamon commented on homeassistant/components/irm_kmi/config_flow.py:
```diff
@@ -0,0 +1,147 @@
+    async def async_step_init(self, user_input: dict | None = None) -> ConfigFlowResult:
+        """Manage the options."""
+        if user_input is not None:
+            _LOGGER.debug(user_input)
```
```suggestion
            _LOGGER.debug("Provided config user is: %s", user_input)
```

jdejaegh replied:
Thanks for the suggested changes.  I implemented them in commit e7af3bbf30e8b3e1a0a4f62be5ff09af3347e9fb

silamon commented on homeassistant/components/irm_kmi/coordinator.py:
```diff
@@ -0,0 +1,134 @@
+            manufacturer=IRM_KMI_NAME.get(
+                preferred_language(self.hass, self.config_entry)
+            ),
+            name=f"{entry.title}",
```
```suggestion
            name=entry.title,
```

silamon commented on homeassistant/components/irm_kmi/coordinator.py:
```diff
@@ -0,0 +1,134 @@
+            # Polling interval. Will only be polled if there are subscribers.
+            update_interval=timedelta(minutes=7),
+        )
+        self.config_entry: ConfigEntry = entry
```
This is already done as part of the TimestampDataUpdateCoordinator initialization. 

```suggestion
```

silamon commented on homeassistant/components/irm_kmi/coordinator.py:
```diff
@@ -0,0 +1,134 @@
+        if self._api.get_city() in OUT_OF_BENELUX:
+            _LOGGER.error(
+                "The zone %s is now out of Benelux and forecast is only available in Benelux. "
+                "Associated device is now disabled.  Move the zone back in Benelux and re-enable to fix "
```
```suggestion
                "Associated device is now disabled. Move the zone back in Benelux and re-enable to fix "
```

jdejaegh committed:
Implement suggested changes

https://github.com/home-assistant/core/pull/144689#pullrequestreview-2848976808

joostlek commented on homeassistant/components/irm_kmi/types.py:
```diff
@@ -0,0 +1,32 @@
+type IrmKmiConfigEntry = ConfigEntry[RuntimeData]
+
+
+class RuntimeData:
+    """Data class for configuration entry runtime data."""
+
+    api_client: IrmKmiApiClientHa
+    coordinator: IrmKmiCoordinator
+
+    def __init__(self, hass: HomeAssistant, entry: IrmKmiConfigEntry) -> None:
+        """Construct RuntimeData using the config entry."""
+
+        api_client = IrmKmiApiClientHa(
+            session=async_get_clientsession(hass),
+            user_agent=USER_AGENT,
+            cdt_map=IRM_KMI_TO_HA_CONDITION_MAP,
+        )
+
+        self.api_client = api_client
+        # If I don't put the api_client in the coordinator this way, I get circular dependencies.
+        self.coordinator = IrmKmiCoordinator(hass, entry, api_client)
```
I think it's best for the runtime data object to be a plain object without logic. It can be a dataclass

jdejaegh replied:
Thanks for your comments.  I'll have a look at them and implement the changes.  
This one is fixed in commit 4c4b9e29927df2a6544a3fbd501f09faed607a5b

joostlek commented on homeassistant/components/irm_kmi/__init__.py:
```diff
@@ -0,0 +1,39 @@
+from homeassistant.exceptions import ConfigEntryError
+
+from .const import PLATFORMS
+from .types import IrmKmiConfigEntry, RuntimeData
```
I would also rename it to something more like `IrmKmiData` or something, RuntimeData sounds very generic

jdejaegh replied:
Renamed in commit 4c4b9e29927df2a6544a3fbd501f09faed607a5b

joostlek commented on homeassistant/components/irm_kmi/config_flow.py:
```diff
@@ -0,0 +1,147 @@
+class IrmKmiConfigFlow(ConfigFlow, domain=DOMAIN):
+    """Configuration flow for the IRM KMI integration."""
+
+    VERSION = CONFIG_FLOW_VERSION
```
I would not use a variable for the version imo

jdejaegh replied:
Constant removed and assigned the literal value directly in commit 895eb40104955a96e82a58580f0416dd2043d1e7

joostlek commented on homeassistant/components/irm_kmi/config_flow.py:
```diff
@@ -0,0 +1,147 @@
+                assert zone is not None  # Assert is here for mypy linting.
+                api_data = {}
+                try:
+                    async with asyncio.timeout(60):
```
ideally timeout is part of the library

jdejaegh replied:
It is already done by the library but somehow I left them here as well.  Fixed in commit 8ece5680f243ce0a20e75b20d3e68cd3e82e07b7

joostlek commented on homeassistant/components/irm_kmi/config_flow.py:
```diff
@@ -0,0 +1,147 @@
+                                "long": zone.attributes[ATTR_LONGITUDE],
+                            }
+                        )
+                except Exception:  # noqa: BLE001
+                    errors["base"] = "api_error"
```
Please log on `.exception` once this happens, because if this happens that means we should be more specific

jdejaegh replied:
Logging added in commit 83489bcd50e6878877a6c514a87a028a9546cdc3

joostlek commented on homeassistant/components/irm_kmi/config_flow.py:
```diff
@@ -0,0 +1,147 @@
+        if user_input:
+            _LOGGER.debug("Provided config user is: %s", user_input)
+
+            if (zone := self.hass.states.get(user_input[CONF_ZONE])) is None:
+                errors[CONF_ZONE] = "zone_not_exist"
```
why do we use a zone instead of lang lot?

jdejaegh replied:
I assume you meant "instead of latitude longitude".  Most of the time, users want weather for their home (which they already configured when they set up Home Assistant) so it seemed natural to pick a zone.  If they want weather for another location, they can create another zone and set up the integration with that one.  

Do you think it should be changed to latitude and longitude in the configuration flow?

silamon replied:
You can do it like other weather integrations and use a LocationSelector with default position set to the home location: https://github.com/home-assistant/core/blob/6de2258325bfb22ef27d2da1f8d6a0cac607227b/homeassistant/components/smhi/config_flow.py#L67-L77
That will lower the complexity as you don't have moving zones anymore and therefore no repair issues.

jdejaegh replied:
Indeed, this would remove the need for repairs.  I'll look into it and implement the configuration flow using latitude and longitude instead of a zone.  Thanks for the pointer to another component

jdejaegh replied:
This is implemented in commit 1ea8bb0.

joostlek commented on homeassistant/components/irm_kmi/config_flow.py:
```diff
@@ -0,0 +1,147 @@
+                    vol.Required(CONF_ZONE): EntitySelector(
+                        EntitySelectorConfig(domain=ZONE_DOMAIN)
+                    ),
+                    vol.Required(
+                        CONF_LANGUAGE_OVERRIDE, default="none"
+                    ): SelectSelector(
+                        SelectSelectorConfig(
+                            options=CONF_LANGUAGE_OVERRIDE_OPTIONS,
+                            mode=SelectSelectorMode.DROPDOWN,
+                            translation_key=CONF_LANGUAGE_OVERRIDE,
+                        )
+                    ),
```
config flow should contain only things needed to configure the integration, no preferences, I would keep it options flow exclusive

jdejaegh replied:
Done in commit dc1daa8d2cde220e380b347fb464da8a7b669748

joostlek commented on homeassistant/components/irm_kmi/config_flow.py:
```diff
@@ -0,0 +1,147 @@
+class IrmKmiOptionFlow(OptionsFlow):
+    """Option flow for the IRM KMI integration, help change the options once the integration was configured."""
+
+    def __init__(self, config_entry: IrmKmiConfigEntry) -> None:
+        """Initialize options flow."""
+        self.current_config_entry = config_entry
+
```
I believe you don't have to input the config entry anymore, please check other implementations

jdejaegh replied:
Indeed, `self.config_entry` is set by the parent class (`OptionsFlow`). I checked with [ibeacon](https://github.com/home-assistant/core/blob/e09dde2ea92c062dc86f142e15d388c9c0a94f98/homeassistant/components/ibeacon/config_flow.py#L50).  Fixed in commit  955c8159b44ca1ed9dccf3db2871d2cb00e20d3a

joostlek commented on homeassistant/components/irm_kmi/const.py:
```diff
@@ -0,0 +1,109 @@
+
+# Dict to map ('ww', 'dayNight') tuple from IRM KMI to HA conditions.
+IRM_KMI_TO_HA_CONDITION_MAP: Final = {
+    (0, "d"): ATTR_CONDITION_SUNNY,
+    (0, "n"): ATTR_CONDITION_CLEAR_NIGHT,
+    (1, "d"): ATTR_CONDITION_SUNNY,
+    (1, "n"): ATTR_CONDITION_CLEAR_NIGHT,
+    (2, "d"): ATTR_CONDITION_LIGHTNING_RAINY,
+    (2, "n"): ATTR_CONDITION_LIGHTNING_RAINY,
+    (3, "d"): ATTR_CONDITION_PARTLYCLOUDY,
+    (3, "n"): ATTR_CONDITION_PARTLYCLOUDY,
+    (4, "d"): ATTR_CONDITION_POURING,
+    (4, "n"): ATTR_CONDITION_POURING,
+    (5, "d"): ATTR_CONDITION_LIGHTNING_RAINY,
+    (5, "n"): ATTR_CONDITION_LIGHTNING_RAINY,
+    (6, "d"): ATTR_CONDITION_POURING,
+    (6, "n"): ATTR_CONDITION_POURING,
+    (7, "d"): ATTR_CONDITION_LIGHTNING_RAINY,
+    (7, "n"): ATTR_CONDITION_LIGHTNING_RAINY,
+    (8, "d"): ATTR_CONDITION_SNOWY_RAINY,
```
how does this work?

silamon replied:
It's a mapping between Home Assistant and the API. The number refers to the type and is being returned by the API and d and n refer to day and night. The full list is documented on the hacs repository: https://github.com/jdejaegh/irm-kmi-ha

joostlek commented on homeassistant/components/irm_kmi/coordinator.py:
```diff
@@ -0,0 +1,137 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+class IrmKmiCoordinator(TimestampDataUpdateCoordinator):
```
why do we use a timestamp one?

jdejaegh replied:
I use the last success time to keep data if the API is unreachable for a short amount of time.  The integration will retry up to two times to refresh before showing "Unavailable" to the user.  See here https://github.com/home-assistant/core/blob/dc1daa8d2cde220e380b347fb464da8a7b669748/homeassistant/components/irm_kmi/coordinator.py#L76-L91

joostlek commented on homeassistant/components/irm_kmi/coordinator.py:
```diff
@@ -0,0 +1,137 @@
+            # Name of the data. For logging purposes.
+            name="IRM KMI weather",
+            # Polling interval. Will only be polled if there are subscribers.
+            update_interval=timedelta(minutes=7),
```
7 is oddly specific

jdejaegh replied:
Some data from the API is changed every 10 minutes so I set the `update_interval` a bit lower.  It could have been 8 or 9 minutes as well

joostlek commented on homeassistant/components/irm_kmi/coordinator.py:
```diff
@@ -0,0 +1,137 @@
+        )
+        self._api = api_client
+        self._zone = get_config_value(entry, CONF_ZONE)
+        self.shared_device_info = DeviceInfo(
+            entry_type=DeviceEntryType.SERVICE,
+            identifiers={(DOMAIN, entry.entry_id)},
+            manufacturer=IRM_KMI_NAME.get(
+                preferred_language(self.hass, self.config_entry)
+            ),
+            name=entry.title,
+        )
```
rather use a base entity for this

jdejaegh replied:
I looked at [ecobee](https://github.com/home-assistant/core/blob/dev/homeassistant/components/ecobee/entity.py#L17) to see how it was done.  Implemented in commit fbb5e09712a1c99c3ac0aba3f0b9ba3ef661d17f

joostlek commented on homeassistant/components/irm_kmi/quality_scale.yaml:
```diff
@@ -0,0 +1,29 @@
+rules:
+  # Bronze
+  action-setup:
+    status: exempt
+    comment: >
+      No service action implemented in this integration at the moment.
+  appropriate-polling:
+    status: done
+    comment: >
+      Polling interval is set to 7 minutes.
+  brands: done
+  common-modules: done
+  config-flow-test-coverage: done
+  config-flow: done
+  dependency-transparency: done
+  docs-actions:
+    status: exempt
+    comment: >
+      No service action implemented in this integration at the moment.
+  docs-high-level-description: done
+  docs-installation-instructions: done
+  docs-removal-instructions: done
+  entity-event-setup: done
+  entity-unique-id: done
+  has-entity-name: done
+  runtime-data: done
+  test-before-configure: done
+  test-before-setup: done
+  unique-config-entry: done
```
Let's evaluate all rules

jdejaegh replied:
They are now all evaluated.  `quality_scale.yaml` is update accordingly in commit 9cfc6e9e786d4373a32550a2f98c6bedcb980322

More info in comment below https://github.com/home-assistant/core/pull/144689#issuecomment-2892175369

jdejaegh committed:
Make `RuntimeData` a dataclass and rename it to `IrmKmiData`

jdejaegh committed:
Do not use constant for config flow version

jdejaegh committed:
Log on exception in config flow

jdejaegh committed:
Remove language override from the config flow (option flow only now)

jdejaegh committed:
Remove unneeded config_entry variable

jdejaegh committed:
Use BaseEntity class for common attributes

jdejaegh committed:
Remove duplicate timeouts (already done in the library)

jdejaegh committed:
Update quality scale with all rules evaluated

jdejaegh commented:
# Quality scale check-list

## Bronze
- [X] `action-setup` - Service actions are registered in async_setup. **No actions are registered**
- [X] `appropriate-polling` - If it's a polling integration, set an appropriate polling interval. **7 minutes interval** https://github.com/home-assistant/core/blob/04c50331514da7d63bb62240cfd5d4e72eabc31c/homeassistant/components/irm_kmi/coordinator.py#L43
- [X] `brands` - Has branding assets available for the integration. https://github.com/home-assistant/brands/pull/5017
- [X] `common-modules` - Place common patterns in common modules.
- [X] `config-flow-test-coverage` - Full test coverage for the config flow. 

<details>
<summary>Show coverage report as of commit 8ece5680f243ce0a20e75b20d3e68cd3e82e07b7</summary>

```
---------- coverage: platform linux, python 3.13.3-final-0 -----------
Name                                              Stmts   Miss  Cover   Missing
-------------------------------------------------------------------------------
homeassistant/components/irm_kmi/__init__.py         24      1    96%   38
homeassistant/components/irm_kmi/config_flow.py      47      0   100%
homeassistant/components/irm_kmi/const.py            16      0   100%
homeassistant/components/irm_kmi/coordinator.py      44      2    95%   71-75
homeassistant/components/irm_kmi/data.py              8      0   100%
homeassistant/components/irm_kmi/entity.py           11      0   100%
homeassistant/components/irm_kmi/repairs.py          50      2    96%   55, 60
homeassistant/components/irm_kmi/types.py             9      0   100%
homeassistant/components/irm_kmi/utils.py            27      4    85%   24, 39, 49, 62
homeassistant/components/irm_kmi/weather.py          84      7    92%   101, 128, 134, 143-149, 155
-------------------------------------------------------------------------------
TOTAL                                               320     16    95%


Results (3.67s):
      17 passed

```

</details>

- [X] `config-flow` - Integration needs to be able to be set up via the UI
    - [X] Uses `data_description` to give context to fields. 
    https://github.com/home-assistant/core/blob/04c50331514da7d63bb62240cfd5d4e72eabc31c/homeassistant/components/irm_kmi/strings.json#L18-L21 
    https://github.com/home-assistant/core/blob/04c50331514da7d63bb62240cfd5d4e72eabc31c/homeassistant/components/irm_kmi/strings.json#L55-L57
    - [X] Uses `ConfigEntry.data` and `ConfigEntry.options` correctly.  **`ConfigEntry.data` is for initial configuration flow and `ConfigEntry.options` is when the users modifies the options after first setup.  The relevant data is accessed properly** 
    https://github.com/home-assistant/core/blob/04c50331514da7d63bb62240cfd5d4e72eabc31c/homeassistant/components/irm_kmi/utils.py#L45-L50
- [X] `dependency-transparency` - Dependency transparency. **MIT License, hosted on PyPI** https://github.com/jdejaegh/irm-kmi-api
- [X] `docs-actions` - The documentation describes the provided service actions that can be used.  **Only provides common weather actions, but documentation links to those** See this PR for all docs related items https://github.com/home-assistant/home-assistant.io/pull/38978
- [X] `docs-high-level-description` - The documentation includes a high-level description of the integration brand, product, or service
- [X] `docs-installation-instructions` - The documentation provides step-by-step installation instructions for the integration, including, if needed, prerequisites
- [X] `docs-removal-instructions` - The documentation provides removal instructions
- [X] `entity-event-setup` - Entity events are subscribed in the correct lifecycle methods. **No entity subscribe to event**
- [X] `entity-unique-id` - Entities have a unique ID.  **Yes, unique is derived from the zone name which is unique** 
https://github.com/home-assistant/core/blob/04c50331514da7d63bb62240cfd5d4e72eabc31c/homeassistant/components/irm_kmi/config_flow.py#L79
https://github.com/home-assistant/core/blob/04c50331514da7d63bb62240cfd5d4e72eabc31c/homeassistant/components/irm_kmi/weather.py#L48
- [X] `has-entity-name` - Entities use has_entity_name = True 
https://github.com/home-assistant/core/blob/04c50331514da7d63bb62240cfd5d4e72eabc31c/homeassistant/components/irm_kmi/weather.py#L40
- [X] `runtime-data` - Use ConfigEntry.runtime_data to store runtime data. **Coordinator and API client are stored there** 
https://github.com/home-assistant/core/blob/04c50331514da7d63bb62240cfd5d4e72eabc31c/homeassistant/components/irm_kmi/types.py#L15-L21 
https://github.com/home-assistant/core/blob/04c50331514da7d63bb62240cfd5d4e72eabc31c/homeassistant/components/irm_kmi/__init__.py#L17
- [X] `test-before-configure` - Test a connection in the config flow. 
https://github.com/home-assistant/core/blob/04c50331514da7d63bb62240cfd5d4e72eabc31c/homeassistant/components/irm_kmi/config_flow.py#L61-L73
- [X] `test-before-setup` - Check during integration initialization if we are able to set it up correctly. 
https://github.com/home-assistant/core/blob/04c50331514da7d63bb62240cfd5d4e72eabc31c/homeassistant/components/irm_kmi/__init__.py#L19-L24
- [X] `unique-config-entry` - Don't allow the same device or service to be able to be set up twice. **A zone can only be configured once.  It will abort as the zone is used for unique id** 
https://github.com/home-assistant/core/blob/04c50331514da7d63bb62240cfd5d4e72eabc31c/homeassistant/components/irm_kmi/config_flow.py#L79-L80

## Silver
- [X] `action-exceptions` - Service actions raise exceptions when encountering failures **No actions are registered**
- [X] `config-entry-unloading` - Support config entry unloading **Done**
https://github.com/home-assistant/core/blob/8ece5680f243ce0a20e75b20d3e68cd3e82e07b7/homeassistant/components/irm_kmi/__init__.py#L46-L48
- [X] `docs-configuration-parameters` - The documentation describes all integration configuration options https://github.com/home-assistant/home-assistant.io/pull/38978
- [X] `docs-installation-parameters` - The documentation describes all integration installation parameters
- [ ] `entity-unavailable` - Mark entity unavailable if appropriate
- [X] `integration-owner` - Has an integration owner @jdejaegh
- [ ] `log-when-unavailable` - If internet/device/service is unavailable, log once when unavailable and once when back connected
- [ ] `parallel-updates` - Number of parallel updates is specified
- [X] `reauthentication-flow` - Reauthentication needs to be available via the UI **There is no authentication for this integration**
- [ ] `test-coverage` - Above 95% test coverage for all integration modules

## Gold
- [X] `devices` - The integration creates devices **Device info is added as part of the BaseEntity class**
https://github.com/home-assistant/core/blob/8ece5680f243ce0a20e75b20d3e68cd3e82e07b7/homeassistant/components/irm_kmi/entity.py#L18-L28
- [ ] `diagnostics` - Implements diagnostics
- [X] `discovery-update-info` - Integration uses discovery info to update network information **Exempt: The integration does not look for devices on the network.  It uses an online API.**
- [X] `discovery` - Devices can be discovered **Same as previous**
- [X] `docs-data-update` - The documentation describes how data is updated. **Done** https://github.com/home-assistant/home-assistant.io/pull/38978
- [ ] `docs-examples` - The documentation provides automation examples the user can use.
- [X] `docs-known-limitations` - The documentation describes known limitations of the integration (not to be confused with bugs) **Mentioned in the documentation** https://github.com/home-assistant/home-assistant.io/pull/38978
- [X] `docs-supported-devices` - The documentation describes known supported / unsupported devices **Exempt: This rule also does not apply to integrations that don't integrate physical devices.**
- [X] `docs-supported-functions` - The documentation describes the supported functionality, including entities, and platforms **Done** https://github.com/home-assistant/home-assistant.io/pull/38978
- [ ] `docs-troubleshooting` - The documentation provides troubleshooting information
- [X] `docs-use-cases` - The documentation describes use cases to illustrate how this integration can be used **Seems clear from the first sentence of the documentation**
- [X] `dynamic-devices` - Devices added after integration setup
- [ ] `entity-category` - Entities are assigned an appropriate EntityCategory **Current entity does not have a category.  Leaving todo as I plan to implement more entities in future PRs.**
- [ ] `entity-device-class` - Entities use device classes where possible **Only weather entity currently. Leaving todo as I plan to implement more entities in future PRs.**
- [ ] `entity-disabled-by-default` - Integration disables less popular (or noisy) entities
- [ ] `entity-translations` - Entities have translated names
- [ ] `exception-translations` - Exception messages are translatable
- [ ] `icon-translations` - Entities implement icon translations
- [X] `reconfiguration-flow` - Integrations should have a reconfigure flow **Exempt: there is no configuration per se, just a zone to pick**
- [X] `repair-issues` - Repair issues and repair flows are used when user intervention is needed **Repair flow available when the zone is outside of the covered area**
https://github.com/home-assistant/core/blob/8ece5680f243ce0a20e75b20d3e68cd3e82e07b7/homeassistant/components/irm_kmi/coordinator.py#L81-L106
- [X] `stale-devices` - Stale devices are removed **Device cannot become stale**

## Platinum
- [X] `async-dependency` - Dependency is async **`irm-kmi-api` uses async functions where possible (namely, http requests with aiohttp**
- [X] `inject-websession` - The integration dependency supports passing in a websession **`irm-kmi-api` uses aiohttp for requests and need a client to provided when creating an instance of the API client**
https://github.com/home-assistant/core/blob/8ece5680f243ce0a20e75b20d3e68cd3e82e07b7/homeassistant/components/irm_kmi/__init__.py#L19-L25
- [ ] `strict-typing` - Strict typing

jdejaegh commented:
Requested changes were addressed: either implementing the suggested change (and marking conversation as resolved) or explaining the decision.  I left 2 conversations as unresolved with a comment: reviewer's opinion on those two is welcome.

PR is ready for review

jdejaegh commented:
Implementing configuration flow using latitude and longitude instead of a zone to simplify the code.  Converted to draft until this is done

jdejaegh committed:
Use lat lon instead of zone

silamon commented on homeassistant/components/irm_kmi/config_flow.py:
```diff
@@ -0,0 +1,122 @@
+                    ).get_forecasts_coord({"lat": lat, "long": lon})
+                except IrmKmiApiError:
+                    errors["base"] = "api_error"
+                    _LOGGER.exception(
```
When you get to this point, it's nice to log the exception as well since this message just says that an error has happened, without going more in details.

silamon commented on homeassistant/components/irm_kmi/config_flow.py:
```diff
@@ -0,0 +1,122 @@
+            lon: float = user_input[CONF_LOCATION][ATTR_LONGITUDE]
+
+            # Check if location is in Benelux.
+            if not errors:
```
It's impossible here to have errors at this point, this can be simplified thanks to using the location.

silamon commented on homeassistant/components/irm_kmi/config_flow.py:
```diff
@@ -0,0 +1,122 @@
+                    )
+
+                if api_data.get("cityName", None) in OUT_OF_BENELUX:
+                    errors[CONF_ZONE] = "out_of_benelux"
```
This error should now be attached to the location selector to display correctly.

jdejaegh committed:
Remove unnecessary error check

jdejaegh committed:
Show error for correct field in config flow

jdejaegh committed:
Fix test after previous commit

jdejaegh committed:
Remove duplicate data from IrmKmiData

jdejaegh committed:
Remove return False from async_setup_entry

jdejaegh commented on homeassistant/components/irm_kmi/__init__.py:
```diff
@@ -0,0 +1,53 @@
+        api_client=api_client,
+        # If I don't put the api_client in the coordinator this way, I get circular dependencies.
+        coordinator=IrmKmiCoordinator(hass, entry, api_client),
+    )
```
Right. `api_client` is no longer in `IrmKmiData` as of 2d24ad183fbbd5e85d0003a08c13477f37fdfa82

jdejaegh commented on homeassistant/components/irm_kmi/__init__.py:
```diff
@@ -0,0 +1,53 @@
+    except ConfigEntryError:
+        # This happens when the zone is out of Benelux (no forecast available there).
+        # This should be caught by the config flow anyway.
+        return False
```
I looked at other integration to see how they did.  They indeed just call the `async_config_entry_firs_refresh()` method without `try-except`.  Example: https://github.com/home-assistant/core/blob/bcb87cf812d4762467a38297aad583076e974e31/homeassistant/components/hko/__init__.py#L17-L33

I removed the `try-except` in c2ff413213cf296984a54602af12aea6557098e5

jdejaegh committed:
Use typed config entry

jdejaegh commented on homeassistant/components/irm_kmi/__init__.py:
```diff
@@ -0,0 +1,53 @@
+
+
+async def async_reload_entry(hass: HomeAssistant, entry: ConfigEntry) -> None:
+    """Reload config entry."""
```
Done f9b678e8ef5ecd96e3987349ecbdde31c32b19d7

jdejaegh committed:
Remove comments

jdejaegh commented on homeassistant/components/irm_kmi/coordinator.py:
```diff
@@ -0,0 +1,93 @@
+            config_entry=entry,
+            # Name of the data. For logging purposes.
+            name="IRM KMI weather",
+            # Polling interval. Will only be polled if there are subscribers.
```
Done 4bdd9cc79c6c83ec8cf8f6aee8c495749c823fce

jdejaegh committed:
Fix type hint

jdejaegh committed:
Fix IrmKmiBaseEntity

jdejaegh commented on homeassistant/components/irm_kmi/entity.py:
```diff
@@ -0,0 +1,28 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+class IrmKmiBaseEntity(Entity):
```
Right, that's easier than having all the sub-classes also inherit from `CoordinatorEntity`.  Done in 1e89b91ba63fa81a3c86874f4ef06edb5609a826

jdejaegh commented on homeassistant/components/irm_kmi/entity.py:
```diff
@@ -0,0 +1,28 @@
+        """Init base properties for IRM KMI entities."""
+        self._attr_attribution = (
+            "Weather data from the Royal Meteorological Institute of Belgium meteo.be"
+        )
```
Done in 1e89b91ba63fa81a3c86874f4ef06edb5609a826

jdejaegh committed:
Remove unneeded None check

jdejaegh commented:
Done the following:
- Rebased to match `home-assistant/core/dev`
- Addressed reviewers' comments (thanks for your reviews)
- Update documentation PR https://github.com/home-assistant/home-assistant.io/pull/38978

The PR is ready for review

jdejaegh committed:
Update dependency

silamon commented on tests/components/irm_kmi/conftest.py:
```diff
@@ -0,0 +1,134 @@
+        return
+
+
+@pytest.fixture
+def mock_get_forecast_api_error_repair():
+    """Mock a call to IrmKmiApiClient.get_forecasts_coord() so that it raises an error."""
+    with patch(
+        "homeassistant.components.irm_kmi.repairs.IrmKmiApiClient.get_forecasts_coord",
+        side_effet=IrmKmiApiError,
+    ):
+        return
```
No longer used? In the other case you may want to set side_effect instead of side_effet.

jdejaegh replied:
Thanks for checking the unused fixtures.  Indeed, not needed anymore.  Removed in 94aa3ca35db19f6bd33c5ff601d269efba2ff33f

silamon commented on tests/components/irm_kmi/conftest.py:
```diff
@@ -0,0 +1,134 @@
+        yield
+
+
+@pytest.fixture
+def mock_get_forecast_api_error():
+    """Mock a call to IrmKmiApiClient.get_forecasts_coord() so that it raises an error."""
+    with patch(
+        "homeassistant.components.irm_kmi.config_flow.IrmKmiApiClient.get_forecasts_coord",
+        side_effet=IrmKmiApiError,
+    ):
+        return
```
No longer used?

jdejaegh replied:
Still needed but it was raising another error than the expected one.  I fixed this in 94aa3ca35db19f6bd33c5ff601d269efba2ff33f

Needed for `test_config_flow_with_api_error`

silamon commented on tests/components/irm_kmi/conftest.py:
```diff
@@ -0,0 +1,134 @@
+        yield irm_kmi
+
+
+@pytest.fixture
+def mock_irm_kmi_api_repair_in_benelux(
+    request: pytest.FixtureRequest,
+) -> Generator[None, MagicMock]:
+    """Return a mocked IrmKmi api client."""
+    fixture: str = "forecast.json"
+
+    forecast = json.loads(load_fixture(fixture, "irm_kmi"))
+    with patch(
+        "homeassistant.components.irm_kmi.repairs.IrmKmiApiClient", autospec=True
+    ) as irm_kmi_api_mock:
+        irm_kmi = irm_kmi_api_mock.return_value
+        irm_kmi.get_forecasts_coord.return_value = forecast
+        yield irm_kmi
```
We no longer have repairs, so this should simplify the amount of tests.

jdejaegh replied:
Removed in 94aa3ca35db19f6bd33c5ff601d269efba2ff33f

silamon commented on tests/components/irm_kmi/test_coordinator.py:
```diff
@@ -0,0 +1,19 @@
+from tests.common import MockConfigEntry
+
+
+async def test_jules_forgot_to_revert_update_interval_before_pushing(
```
Can you give this a clearer name? We don't need unit tests to blame developers for their bugs. 😉

jdejaegh committed:
Clean test fixtures

jdejaegh committed:
Fix wording in tests

jdejaegh committed:
Make test name and doc more meaningful

joostlek commented on homeassistant/components/irm_kmi/config_flow.py:
```diff
@@ -0,0 +1,120 @@
+
+                name: str = api_data.get("cityName", "")
+                if user_input[CONF_LOCATION] == home_location:
+                    name = HOME_LOCATION_NAME
```
might as well take the instance name, because we can't translate this

jdejaegh replied:
Indeed, removed the condition in 31e15bbdb4094e9403aae8ac4d042de5038b6283

joostlek commented on homeassistant/components/irm_kmi/utils.py:
```diff
@@ -0,0 +1,33 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+def get_config_value(config_entry: ConfigEntry, key: str, default: Any = None) -> Any:
+    """Get the value of key in the configuration.  If options were modified, they take priority."""
+
+    if config_entry.options and key in config_entry.options:
+        return config_entry.options[key]
+    if config_entry.data and key in config_entry.data:
+        return config_entry.data[key]
+    return default
```
I am not sure what the benefit of this is. It's either in options or in data, not in both. So I think we should be specific instead of "Get it from somewhere"

joostlek commented on homeassistant/components/irm_kmi/coordinator.py:
```diff
@@ -0,0 +1,90 @@
+                and self.last_update_success_time - utcnow()
+                < timedelta(seconds=2.5 * self.update_interval.seconds)
+            ):
+                _LOGGER.warning(
+                    "Error communicating with API for general forecast: %s. Keeping the old data",
+                    err,
+                )
+                return self.data
```
when does this happen?

jdejaegh replied:
This happens when the server timeouts or when it is unreachable for some reason.  I have seen this happen a few times in my logs.

joostlek commented on homeassistant/components/irm_kmi/coordinator.py:
```diff
@@ -0,0 +1,90 @@
+
+        return await self.process_api_data()
+
+    async def async_refresh(self) -> None:
+        """Refresh data and log errors."""
+        await self._async_refresh(log_failures=True, raise_on_entry_error=True)
```
why do we have this?

joostlek commented on homeassistant/components/irm_kmi/data.py:
```diff
@@ -0,0 +1,16 @@
+from homeassistant.components.weather import Forecast
+
+
+class ProcessedCoordinatorData(dict[str, Any]):
+    """Data class that will be exposed to the entities consuming data from an IrmKmiCoordinator."""
+
+    current_weather: CurrentWeatherData
+    hourly_forecast: list[Forecast] | None
+    daily_forecast: list[ExtendedForecast] | None
+    country: str
```
why does it extend a `dict`? Should we make this a dataclass?

jdejaegh replied:
By extending `dict` it inherits all the methods of `dict`, like `get()`.  I turned this into a dataclass with `get()` implemented in 8aa7d2ab122

joostlek commented on homeassistant/components/irm_kmi/entity.py:
```diff
@@ -0,0 +1,32 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+class IrmKmiBaseEntity(CoordinatorEntity):
```
```suggestion
class IrmKmiBaseEntity(CoordinatorEntity[IrmKmiCoordinator]):
```

joostlek commented on homeassistant/components/irm_kmi/entity.py:
```diff
@@ -0,0 +1,32 @@
+
+    _attr_attribution = (
+        "Weather data from the Royal Meteorological Institute of Belgium meteo.be"
+    )
```
```suggestion
    )
    _attr_has_entity_name = True
```

joostlek commented on homeassistant/components/irm_kmi/entity.py:
```diff
@@ -0,0 +1,32 @@
+        "Weather data from the Royal Meteorological Institute of Belgium meteo.be"
+    )
+
+    def __init__(self, entry: ConfigEntry, name: str) -> None:
```
use the typed config entry

jdejaegh replied:
Fixed in 31e15bbdb4094e9403aae8ac4d042de5038b6283

joostlek commented on homeassistant/components/irm_kmi/entity.py:
```diff
@@ -0,0 +1,32 @@
+        "Weather data from the Royal Meteorological Institute of Belgium meteo.be"
+    )
+
+    def __init__(self, entry: ConfigEntry, name: str) -> None:
```
what's the name?

silamon replied:
It's fine to pass it to a base entity, given that it will be reused when other platforms are added. You may want to change the hardcoded string to a translatable name by using entity name translations.

jdejaegh replied:
Solved by this discussion: https://github.com/home-assistant/core/pull/144689#discussion_r2265313345

joostlek commented on homeassistant/components/irm_kmi/entity.py:
```diff
@@ -0,0 +1,32 @@
+
+    def __init__(self, entry: ConfigEntry, name: str) -> None:
+        """Init base properties for IRM KMI entities."""
+        coordinator = entry.runtime_data.coordinator
+        CoordinatorEntity.__init__(self, coordinator)
```
Instead also just pass the coordinator on, it's probably a bit cleaner

```suggestion
        super().__init__(coordinator)
```

jdejaegh replied:
Done in 31e15bbdb4094e9403aae8ac4d042de5038b6283

jdejaegh committed:
Address comments on PR

jdejaegh committed:
Make code more specific to get options and config values

jdejaegh committed:
Use dataclass instead of dict

jdejaegh committed:
Address comments in PR

Fix .get() methods

Do not nest runtime data for storing only one value

Use IrmKmiConfigEntry where relevant

jdejaegh commented on homeassistant/components/irm_kmi/coordinator.py:
```diff
@@ -0,0 +1,86 @@
+    """Coordinator to update data from IRM KMI."""
+
+    def __init__(
+        self, hass: HomeAssistant, entry: ConfigEntry, api_client: IrmKmiApiClientHa
```
Fixed in 1028e361aba

jdejaegh commented on homeassistant/components/irm_kmi/types.py:
```diff
@@ -0,0 +1,16 @@
+class IrmKmiData:
+    """Data class for configuration entry runtime data."""
+
+    coordinator: IrmKmiCoordinator
```
Done in 1028e361aba
And moved the `type IrmKmiConfigEntry` to `coordinator.py`

jdejaegh committed:
Remove unneeded code and strings

jdejaegh commented on homeassistant/components/irm_kmi/strings.json:
```diff
@@ -0,0 +1,54 @@
+          "language_override": "[%key:common::config_flow::data::language%]"
+        },
+        "data_description": {
+          "language_override": "[%key:component::irm_kmi::common::language_override_description%]",
```
Removed in 5933c00a65c

jdejaegh commented on homeassistant/components/irm_kmi/strings.json:
```diff
@@ -0,0 +1,54 @@
+    "error": {
+      "out_of_benelux": "The location is outside of Benelux. Pick a location in Benelux.",
+      "api_error": "[%key:common::config_flow::error::cannot_connect%]",
+      "config_entry_none": "Unknown configuration entry"
```
Removed in 5933c00a65c

jdejaegh commented on homeassistant/components/irm_kmi/entity.py:
```diff
@@ -0,0 +1,34 @@
+from .types import IrmKmiConfigEntry
+from .utils import preferred_language
+
+_LOGGER = logging.getLogger(__name__)
```
Removed in 5933c00a65c

jdejaegh committed:
Remove .get() method on ProcessedCoordinatorData

joostlek commented on homeassistant/components/irm_kmi/config_flow.py:
```diff
@@ -0,0 +1,117 @@
+                    "Encountered an unexpected error while configuring the integration"
+                )
+
+            if api_data.get("cityName", None) in OUT_OF_BENELUX:
```
Okay, if we expect it always to be there, use `["cityName"]`. this also solves your next comment

jdejaegh committed:
Remove get_config_value_data and get_config_value_options

jdejaegh committed:
Use SingleCoordinatorWeatherEntity instead of WeatherEntity

jdejaegh commented on homeassistant/components/irm_kmi/weather.py:
```diff
@@ -0,0 +1,170 @@
+    async_add_entities([IrmKmiWeather(entry)])
+
+
+class IrmKmiWeather(WeatherEntity, IrmKmiBaseEntity):
```
I did not explore that option before reading your comment.  The class you refer to is probably `CoordinatorWeatherEntity`.  This class can take multiple coordinators for different types of forecasts (a bit overkill in this case, I have only one coordinator).  There is a sub-class that allows taking only one coordinator for all forecasts `SingleCoordinatorWeatherEntity`

As the changes to use that class instead of the basic `WeatherEntity` were minimal, I implemented that in 56c16cae787

jdejaegh committed:
Remove 'Forecast' to be set as DeviceInfo name, use config entry title instead

jdejaegh commented on homeassistant/components/irm_kmi/weather.py:
```diff
@@ -0,0 +1,170 @@
+    def __init__(self, entry: IrmKmiConfigEntry) -> None:
+        """Create a new instance of the weather entity from a configuration entry."""
+        WeatherEntity.__init__(self)
+        IrmKmiBaseEntity.__init__(self, entry, "Forecast")
```
That's indeed easier than having a fixed string.  Done in 4c8a42d4974

jdejaegh committed:
Use shorthand attributes

jdejaegh committed:
Make current_weather a property

jdejaegh committed:
Use OptionsFlowWithReload

jdejaegh commented on homeassistant/components/irm_kmi/__init__.py:
```diff
@@ -0,0 +1,44 @@
+    await entry.runtime_data.coordinator.async_config_entry_first_refresh()
+
+    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+    entry.async_on_unload(entry.add_update_listener(async_reload_entry))
```
It is now using `OptionsFlowWithReload` in 8f87d2ca0a8

jdejaegh committed:
Prevent adding twice the same city

jdejaegh commented on homeassistant/components/irm_kmi/weather.py:
```diff
@@ -0,0 +1,170 @@
+        WeatherEntity.__init__(self)
+        IrmKmiBaseEntity.__init__(self, entry, "Forecast")
+        self._name = entry.title
+        self._attr_unique_id = entry.entry_id
```
The best unique id I can think of is the name of the city for which the forecasts are given.  The API does some geocoding from the location to give the name of municipality.  Having twice the same municipality set up is indeed useless.  There is not weather station concept in the API.  You can only request data for a location.

I used the city and country code (lower case) as unique id (e.g. `leuven be` or `namur be`).  This prevents to add twice the same city, even if you move slightly the pin on the map.

Implemented in a15c633a111

joostlek replied:
This hasn't changed yet

jdejaegh replied:
My mistake, I only changed it in the config flow and forgot weather entity.  Done in dc348ac62a6d35d6ef7898b4a96338b0fd2bb20d

jdejaegh commented on homeassistant/components/irm_kmi/strings.json:
```diff
@@ -0,0 +1,54 @@
+      "config_entry_none": "Unknown configuration entry"
+    }
+  },
+  "selector": {
+    "language_override": {
+      "options": {
+        "none": "Follow Home Assistant server language",
+        "fr": "French",
+        "nl": "Dutch",
+        "de": "German",
+        "en": "English"
+      }
+    }
+  },
```
It would but it's not usable due to the "Follow Home Assistant server language" option.  I just tried with `LanguageSelector` but it does not seem possible to add options that are not languages.  In my opinion, the current selector fits the use-case and is not too hard to read compared to a language selector.  

The only downside of the current setup is that you cannot have the names automatically translated.  I could go in `strings.json` and change the strings to have the native name of each language (Français instead of French).  Would that be a better solution?

NoRi2909 replied:
If a user decides to run Home Assistant with the English UI they will have certainly no problem with the language names listed in English.

Anyone who runs Home Assistant in French, Dutch or German will see these in their preferred languages as well as we will translate these strings in Lokalise.

jdejaegh committed:
Remove unused logger

jdejaegh commented:
Thanks for the comments 

I addressed the comments and closed them when fully addressed.  Two discussions are still open and feedback is welcome:
- https://github.com/home-assistant/core/pull/144689#discussion_r2267609906 about what to do if current weather were to go missing
- https://github.com/home-assistant/core/pull/144689#discussion_r2267710655 about language selector.  Given the feedback, I would stick with the current implementation and names in English in `strings.json`

This PR is ready for review

joostlek commented:
Double check this one btw https://github.com/home-assistant/core/pull/144689#discussion_r2267451906

jdejaegh committed:
Use [] instead of get

jdejaegh committed:
Use [] instead of get

joostlek commented on tests/components/irm_kmi/test_config_flow.py:
```diff
@@ -0,0 +1,118 @@
+    }
+
+
+async def test_user_flow_home(
+    hass: HomeAssistant,
+    mock_setup_entry: MagicMock,
+    mock_get_forecast_in_benelux: MagicMock,
+) -> None:
+    """Test the full user configuration flow."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": SOURCE_USER}
+    )
+    hass.config.latitude = 50.123
+    hass.config.longitude = 4.456
+
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"],
+        user_input={CONF_LOCATION: {ATTR_LATITUDE: 50.123, ATTR_LONGITUDE: 4.456}},
+    )
+    assert result.get("type") == FlowResultType.CREATE_ENTRY
+    assert result.get("title") == "Brussels"
```
What does this test?

jdejaegh replied:
This checks that the config entry title is set to the city name returned by the weather provider.  This was used for the name of the weather entity (which has been set to `None` after comments). 
The config entry still needs a title so the city name is used.

joostlek commented on tests/components/irm_kmi/test_config_flow.py:
```diff
@@ -0,0 +1,118 @@
+    assert result2.get("type") == FlowResultType.FORM
+    assert result2.get("step_id") == "user"
+    assert "base" in result2.get("errors")
+
```
I am missing a test where we try to set up the same location twice

joostlek commented on tests/components/irm_kmi/test_config_flow.py:
```diff
@@ -0,0 +1,118 @@
+        result["flow_id"], user_input={}
+    )
+
+    assert result["type"] == FlowResultType.CREATE_ENTRY
```
```suggestion
    assert result["type"] is FlowResultType.CREATE_ENTRY
```

joostlek commented on tests/components/irm_kmi/test_coordinator.py:
```diff
@@ -0,0 +1,23 @@
+from tests.common import MockConfigEntry
+
+
+def test_update_interval_is_7_minutes(
+    hass: HomeAssistant,
+    mock_config_entry: MockConfigEntry,
+) -> None:
+    """Test that the refresh interval 7 minutes.
+
+    If you need to change this test, also change the documentation: https://github.com/home-assistant/home-assistant.io.
+    """
+
+    coordinator = IrmKmiCoordinator(hass, mock_config_entry, MagicMock())
```
Meh, generally we don't have tests for this. If you want to implicitly test this, add a test to the weather platform where you update, freeze time and skip 7 minutes and observe it updated again

jdejaegh replied:
Got it!  I used this test as a safeguard when developing to avoid pushing code with a low refresh interval.  I guess it is not needed anymore.  Removed it in 54c7745d515

joostlek commented on tests/components/irm_kmi/test_init.py:
```diff
@@ -0,0 +1,58 @@
+    mock_irm_kmi_api: AsyncMock,
+) -> None:
+    """Test the IRM KMI configuration entry loading/unloading."""
+    mock_config_entry.runtime_data.api_client = mock_irm_kmi_api
```
we should not touch runtime_data

joostlek commented on tests/components/irm_kmi/test_init.py:
```diff
@@ -0,0 +1,58 @@
+    """Test the IRM KMI configuration entry loading/unloading."""
+    mock_config_entry.runtime_data.api_client = mock_irm_kmi_api
+
+    hass.states.async_set(
+        "zone.home",
+        0,
+        {"latitude": 50.738, "longitude": 4.054},
+    )
+
+    mock_config_entry.add_to_hass(hass)
+
+    await hass.config_entries.async_setup(mock_config_entry.entry_id)
+    await hass.async_block_till_done()
+
+    assert mock_config_entry.state is ConfigEntryState.LOADED
+
+    await hass.config_entries.async_unload(mock_config_entry.entry_id)
+    await hass.async_block_till_done()
+
+    assert not hass.data.get(DOMAIN)
+    assert mock_config_entry.state is ConfigEntryState.NOT_LOADED
+
+
+async def test_config_entry_not_ready(
+    hass: HomeAssistant,
+    mock_config_entry: MockConfigEntry,
+    mock_exception_irm_kmi_api: AsyncMock,
+) -> None:
+    """Test the IRM KMI configuration entry not ready."""
+    hass.states.async_set(
+        "zone.home",
+        0,
+        {"latitude": 50.738681639, "longitude": 4.054077148},
+    )
```
not sure what the home zone is used for

jdejaegh replied:
Not used anymore. It was needed when the zone was used to get the location.  Removed in 7486b3ad914

joostlek commented on tests/components/irm_kmi/test_weather.py:
```diff
@@ -0,0 +1,73 @@
+from tests.common import MockConfigEntry, async_load_fixture
+
+
+@freeze_time(datetime.fromisoformat("2023-12-28T15:30:00+01:00"))
+async def test_weather_nl(
+    hass: HomeAssistant, mock_config_entry: MockConfigEntry
+) -> None:
+    """Test weather with forecast from the Netherland."""
+    mock_config_entry.runtime_data = MagicMock()
+    mock_config_entry.runtime_data.api_client = IrmKmiApiClientHa(MagicMock(), "test")
+    forecast = json.loads(await async_load_fixture(hass, "forecast_nl.json", "irm_kmi"))
+    mock_config_entry.runtime_data.api_client._api_data = forecast
+
+    mock_config_entry.runtime_data = coordinator = IrmKmiCoordinator(
+        hass, mock_config_entry, mock_config_entry.runtime_data.api_client
+    )
+    coordinator.data = await coordinator.process_api_data()
+
+    weather = IrmKmiWeather(mock_config_entry)
+    result = await weather.async_forecast_daily()
+
+    assert isinstance(result, list)
+    assert len(result) == 7
+
+    # When getting daily forecast, the min temperature of the current day
+    # should be the min temperature of the coming night
+    assert result[0]["native_templow"] == 9
```
Okay so in `test_init.py` you did a good job, we should have HA set up the mocked config entry and then we can observe the state machine for the created entities

joostlek reviewed (changes requested):
I think the PR is almost there, feel free to send me a message on Discord so we can get it over the finish line

jdejaegh committed:
Unique id for Weather entity

jdejaegh committed:
Remove getter for name

jdejaegh commented on homeassistant/components/irm_kmi/weather.py:
```diff
@@ -0,0 +1,154 @@
+    @property
+    def name(self) -> str:
+        """Return the name of the entity."""
+        return self._name
```
Removed in dbf3e1d83a0

jdejaegh committed:
Reuse result variable and use is instead of ==

jdejaegh commented on tests/components/irm_kmi/test_config_flow.py:
```diff
@@ -0,0 +1,118 @@
+
+    assert result2.get("type") == FlowResultType.FORM
+    assert result2.get("step_id") == "user"
+    assert "base" in result2.get("errors")
```
Re-used `result` everywhere in `test_config_flow.py` bf09e6d46d1

jdejaegh committed:
Do not manually assign runtime_data in tests

jdejaegh commented on tests/components/irm_kmi/conftest.py:
```diff
@@ -0,0 +1,90 @@
+        data={CONF_LOCATION: {ATTR_LATITUDE: 50.84, ATTR_LONGITUDE: 4.35}},
+        unique_id="50.84-4.35",
+    )
+    entry.runtime_data = MagicMock()
```
Removed manual assignation in 57b1dc73be5

jdejaegh committed:
Remove update interval test

jdejaegh committed:
Abort config on api error + test flow result everytime

jdejaegh committed:
Add test: setup twice same location

jdejaegh committed:
Remove unneeded setup in test

jdejaegh committed:
Rewrite tests

jdejaegh committed:
Use hass.data to get the weather entity in tests

jdejaegh committed:
Mark unavailable when API is unreachable + log

jdejaegh committed:
Set numer of parallel updates

jdejaegh committed:
Improve test_init.py

jdejaegh commented:
Thanks for the comments.  Here is a summary of the changes:
- I addressed the comments and updated the code accordingly
- I checked the quality scale and updated the criteria if needed
- I rebased this branch on core/dev

I have been a bit busy offline in past few weeks, hence the delay before I could look into the PR. 

I'll also have a look at the documentation PR to see if everything is up-to-date with the latest changes.

This PR is ready for review

joostlek committed:
Fix

joostlek reviewed (approved):
Okay I updated the weather tests as the beta cut is nearing. The issue was that we shouldn't access entities directly either in the code or in the tests after creating it. I now added snapshot tests which can be updated using `pytest ./tests/components/irm_kmi --snapshot-update`. These fixate the data we put into them.

Mind sending me a message on Discord so we can add you to the relevant channels on the discord?
